### DISCUSSION TITLE
RAN metric data inside the RIC indication messages

### DISCUSSION THREAD

**Discussion by yc541** [srsran/srsRAN_Project#295] [General Help] [answered]
Hi,

I noticed that the RIC/xApp tutorial [https://docs.srsran.com/projects/project/en/latest/tutorials/source/flexric/source/index.html](url) has updated with some new metrics, well done! The tutorial mentions that the requested metric is hardcoded in the RIC and the xApp can't easily access it. I did some slight modifications to the RIC so the xApp can choose what metric to request for (but still only one metric per xApp). 

Back to the title, within the RIC indication messages from the RAN, how do I identify which RAN node / UE is the decoded metric for? E.g., I have one gnb and two connected UEs, and an xApp is requesting the CQI from the gnb. One UE has good CQI and the other has bad CQI so I can tell which UE corresponds to the reported CQI. The xApp is only getting the CQI of one of the UEs and once I turn this UE off the xApp starts to get the CQI of the other UE. Also is it possible to identify which RAN node is the RIC indication message from (when there are multiple gnbs)?

Attached is a pcap when the xApp is requesting UL throughput while 2 UEs are connected to the gnb. 
[2UE.zip](https://github.com/srsran/srsRAN_Project/files/13283643/2UE.zip)

**Comment by fllay** [comment]
Could you share your modification to the RIC?

**Comment by yc541** [comment]
> Could you share your modification to the RIC?

I just used the enum (e.g., ms_5) where the xApp specifies the KPM indication interval, which is chosen by the xApp and gets passed to the RIC. The RIC then builds the subscription request based on the enum from the xApp which also indicates the metric string it wants.

**Reply by fllay** [reply]
Please let me know which file you can edit the metric string. I think it is in this file 
`src/sm/kpm_sm_v2.02/kpm_sm_ric.c` where  
```
  // Action Definition IE
  long noLabelnum = 0; 
        adapter_LabelInfoItem_t dummylabelInfo = {.noLabel = &noLabelnum}; 
  MeasInfo_t dummy_MeasInfo = { 
    .meas_type = KPM_V2_MEASUREMENT_TYPE_NAME, 
    .measName = {.buf = (unsigned char *)"DRB.RlcSduTransmittedVolumeDL", .len = 29},
    .labelInfo = &dummylabelInfo, 
    .labelInfo_len = 1
  };
  kpm_action_def_t action_def = { 
    .ric_style_type = 1, 
    .granularity_period = 1,
    .MeasInfo = &dummy_MeasInfo,
    .MeasInfo_len = 1, 
    .cell_global_id = KPMV2_CELL_ID_CHOICE_NOTHING
  };
```
I modified this ` .measName = {.buf = (unsigned char *)"DRB.RlcSduTransmittedVolumeDL", .len = 29},` to ` .measName = {.buf = (unsigned char *)"CQI", .len = 3},`. It still reports RSRP metric. BTW, I am using RIC with real UE ( a phone). Did it work only ZMQ configuration?

**Reply by yc541** [reply]
It's the right place to modify, and use e2sm_kpm_meas_provider_test.cpp as a guide to what metric string to put in. And like @yagoda said FlexRIC needs to be rebuilt after this. And I am using COTS handset as well.

**Comment by yagoda** [comment]
Hey,

I think you need to rebuild the whole flexric library with sudo make
install etc in order for your change to take effect.

**Comment by pgawlowicz** [comment]
Hi, the Flexric version that we use here sends only E2SM_KPM Report Service Style 1 when sending an RIC Subscription Request. Report Service Style 1 is for E2-Level measurements (e.g. total DL throughput) and does not contain any UE identifier. Note that our E2 agent already supports all Report Styles (1-5), and styles 2-5 are for UE-level measurements (eg. UE DL throughput).

Currently, our code allows you to request the CQI metric with Report Service Style 1, but this option is used mostly for testing and will be removed soon. It does not check any UE identifier (as there are none in the Report Service Style 1) and simply returns CQI of a UE that is the first in the metric  array -- see [here](https://github.com/srsran/srsRAN_Project/blob/main/lib/e2/e2sm/e2sm_kpm/e2sm_kpm_du_meas_provider_impl.cpp#L220)

**Reply by fllay** [reply]
I did rebuild everything. I remove build directory and cmake it again. I even edit xApp and put special characters in printf in the call back function and it did prints those characters.  Only one difference is that I use the actual device i.e. B210 and a phone. However,  the reported RSRP seems to match with the RSRP in gnb log console. Anyway, I will recheck everything again and report back.

**Reply by yc541** [reply]
Thanks for the explanation, I noticed this yesterday after reading the SRS code you mentioned and the FlexRIC code. The FlexRIC mir-dev and kpm branches seem to have the updated kpm_sm_v03.00 library which supports all 5 report styles, but modifications need to be made to get the Flexric to talk to SRS. The code structure is different from the kpm_sm_v02.00 library and the kpm_sm_v03.00 library in these two branches are not exactly the same. 

Also I noticed that the reported RSRP is actually pusch SNR, but I guess it will be removed at some point since it's just a dummy metric.

**Reply by fllay** [reply]
I got it working now. I for got to `make install`. Sorry, how dumb I am.

**Comment by yc541** [comment]
Hi @pgawlowicz,

I did a bit more experiments with the Flexric mir_dev branch which has all 5 RIC styles implemented. The E2 setup is successful but the gnb has trouble decoding the RIC subscription request from the kpm monitoring xapp. More specifically, the asn1 decoder has issues unpacking E2SM KPM Action Definition, and the unpacked action definition format does not match the action definition format configured in the xapp. Here are the modifications I made to the gnb and Flexric code so the error can be reproduced:

Changes to gnb:
- Change _e2sm_kpm_asn1_packer::ran_func_id_ in _e2sm_kpm_asn1_packer.cpp_ from 147 to 2, which is the KPM SM ID that Flexric accepts
- In the code block not far below the above section:
   printf("Failed to unpack E2SM KPM Action Definition\n");               
   _variant_get<e2_sm_kpm_action_definition_s>(action_def.action_definition).to_json(js);
   std::cout << "AD PDU: " << js.to_string();_
   The last two lines print out the action definition PDU

Changes to Flexric:
- In _CMakeLists_ in the flexric root directory, in the line _set(KPM_VERSION "KPM_V2_01" CACHE STRING "The KPM SM version to use")_, replace "KPM_V2_01" with "KPM_V3_00" to use the kpm library which supports all 5 RIC styles
- In [xapp_kpm_moni.c](https://gitlab.eurecom.fr/mosaic5g/flexric/-/blob/mir_dev/examples/xApp/c/monitor/xapp_kpm_moni.c), in the line below _case ngran_gNB: ;_, replace the line with _const char *act_gnb[] = {"DRB.RlcPacketDropRateDl", "DRB.UEThpUl", NULL};_  which are the metrics SRS supports 

In [xapp_kpm_moni.c](https://gitlab.eurecom.fr/mosaic5g/flexric/-/blob/mir_dev/examples/xApp/c/monitor/xapp_kpm_moni.c) , inside the code block _kpm_act_def_t gen_act_def(const char** act)_  the action definition format is set to 4.

The gnb logs, gnb terminal outputs and pcap are attached.

In the gnb outputs, the ric style is correctly decoded as 4, but the action definition format is 2, which probably caused the subsequent decoding errors in the gnb logs (e.g., meas_type_c)

[SRS_Flexric_logs.zip](https://github.com/srsran/srsRAN_Project/files/13352275/SRS_Flexric_logs.zip)

**Reply by pgawlowicz** [reply, marked as answer]
Hi @yc541,
Thank you for the investigation and your feedback. 
Indeed, the newest FlexRIC version provides much more functionality.

We are currectly working on the example xApp on top of FlexRIC. We have already fixed some small bugs (including the one you pointed). Most probably all will be available after the next code release, along with the updated tutorial in our docs.

Best,
Piotr

**Reply by yc541** [reply]
Thanks for the update, look forward to testing the next SRS release.

**Reply by mas-oliveira** [reply]
Hello @pgawlowicz,

Is somewhere documented the way that we can change these files to receive the metrics available but using the e2ap-v2 branch of FlexRIC?

**Reply by pgawlowicz** [reply]
@mas-oliveira Please check our updated [tutorial](https://docs.srsran.com/projects/project/en/latest/tutorials/source/flexric/source/index.html), where we use the `br-flexric` branch that allows specifying metric names using a config file that is passed to a xApp.

**Reply by SyedDanialAliShah** [reply]
Hi,

The link to the updated tutorial is not working for me? Could i please get a working link? 

Thank u

**Reply by pgawlowicz** [reply]
https://docs.srsran.com/projects/project/en/latest/tutorials/source/near-rt-ric/source/index.html

**Comment by yagoda** [comment]
Hi @yc541,

In preparation for our upcoming workshop in DC, we are contacting some users of our E2 solution. Can you please share a few lines about your experience with the E2 interface, what you were trying to achieve, and how it worked out for you?

Kind Regards,
Justin
